pkg/statedb/txn.go

```diff
@@ -300,41 +318,67 @@ func (txn *txn) Delete(meta TableMeta, guardRevision Revision, data any) (any, b
 	// revert the change.
 	if guardRevision > 0 {
 		if obj.revision != guardRevision {
-			idIndexTree.Insert(idKey, obj)
+			idIndexTree.txn.Insert(idKey, obj)
 			table.revision = oldRevision
 			return obj, true, ErrRevisionNotEqual
 		}
 	}
 
 	// Update revision index.
 	indexTree := txn.mustIndexWriteTxn(tableName, RevisionIndex)
-	if _, ok := indexTree.Delete(revisionKey(obj.revision, idKey)); !ok {
+	if _, ok := indexTree.txn.Delete(revisionKey(obj.revision, idKey)); !ok {
 		panic("BUG: Object to be deleted not found from revision index")
 	}
 
 	// Then update secondary indexes.
 	for idx, indexer := range meta.secondaryIndexers() {
 		indexer.fromObject(obj).Foreach(func(key index.Key) {
 			if !indexer.unique {
-				key = append(key, idKey...)
+				key = encodeMultiIndexKey(idKey, key)
 			}
-			txn.mustIndexWriteTxn(tableName, idx).Delete(key)
+			txn.mustIndexWriteTxn(tableName, idx).txn.Delete(key)
 		})
 	}
 
 	// And finally insert the object into the graveyard.
 	if txn.hasDeleteTrackers(tableName) {
 		graveyardIndex := txn.mustIndexWriteTxn(tableName, GraveyardIndex)
 		obj.revision = revision
-		if _, existed := graveyardIndex.Insert(idKey, obj); existed {
+		if _, existed := graveyardIndex.txn.Insert(idKey, obj); existed {
 			panic("BUG: Double deletion! Deleted object already existed in graveyard")
 		}
-		txn.mustIndexWriteTxn(tableName, GraveyardRevisionIndex).Insert(revisionKey(revision, idKey), obj)
+		txn.mustIndexWriteTxn(tableName, GraveyardRevisionIndex).txn.Insert(revisionKey(revision, idKey), obj)
 	}
 
 	return obj.data, true, nil
 }
 
+// encodeMultiIndexKey constructs the internal key to use with multi-indexes (non-unique).
+// It concatenates the secondary key with the primary key and the length of the secondary key.
+// The length is stored as unsigned 16-bit big endian.
+// This allows looking up from the multi-index with the secondary key by doing a prefix
+// search. The length is used to safe-guard against indexers that don't terminate the key
+// properly (e.g. if secondary key is "foo", then we don't want "foobar" to match).
+func encodeMultiIndexKey(primary, secondary []byte) []byte {
```

Comment:
applies everywhere: `multi` -> `dual` or `duo` or `secondary`

I think multi is mildly misleading, as these functions don't take more than two indices. (I believe the construction could be extended to support more indices, but I don't see why we'd want to. Uniqueness is guaranteed by the inclusion of the primary index.)

Reply:
This is the key for multi-index (non-unique), hence `MultiIndexKey`, it isn't about the number of indices. Could call it `NonUniqueIndexKey`?

Reply:
ah, I see - works for me!